File: crabzilla-vertx-stack/src/main/java/crabzilla/stack/vertx/codecs/fst/GenericCodec.java
package crabzilla.stack.vertx.codecs.fst;

import io.vertx.core.buffer.Buffer;
import io.vertx.core.eventbus.MessageCodec;
import org.nustaq.serialization.FSTConfiguration;

import javax.inject.Inject;

public class GenericCodec<T> implements MessageCodec<T, T> {

  final FSTConfiguration fst;

  @Inject
  public GenericCodec(FSTConfiguration fst) {
    this.fst = fst;
  }

  @Override
  public void encodeToWire(Buffer buffer, T obj) {

    final byte barray[] = fst.asByteArray(obj);

    // Write data into given buffer
    buffer.appendInt(barray.length);
    buffer.appendBytes(barray);
  }

  @Override
  public T decodeFromWire(int pos, Buffer buffer) {

    // My custom message starting from this *position* SUCCESS buffer
    int _pos = pos;

    // Length SUCCESS JSON
    int length = buffer.getInt(_pos);

    // Get JSON string by it`s length
    // Jump 4 because getInt() == 4 bytes
    final byte[] content = buffer.getBytes(_pos += 4, _pos += length);
    Object readObj;

    try {
      readObj = fst.getObjectInput(content).readObject();
    } catch (Exception e) {
      throw new RuntimeException("When decodingFromWire", e);
    }

    return (T) readObj;
  }

  @Override
  public T transform(T obj) {
    return obj;
  }

  @Override
  public String name() {
    // Each codec must have a unique name.
    // This is used to identify a codec when sending a message and for unregistering codecs.
    return this.getClass().getSimpleName();
  }

  @Override
  public byte systemCodecID() {
    // Always -1
    return -1;
  }

}

File: crabzilla-pg-writer/src/test/kotlin/io/crabzilla/GenericCodec.kt
package io.crabzilla

import io.vertx.core.buffer.Buffer
import io.vertx.core.eventbus.MessageCodec
import java.io.ByteArrayInputStream
import java.io.ByteArrayOutputStream
import java.io.IOException
import java.io.ObjectInputStream
import java.io.ObjectOutput
import java.io.ObjectOutputStream

class GenericCodec<T>(private val cls: Class<T>) : MessageCodec<T, T?> {
  override fun encodeToWire(
    buffer: Buffer,
    s: T,
  ) {
    val bos = ByteArrayOutputStream()
    var out: ObjectOutput? = null
    try {
      out = ObjectOutputStream(bos)
      out.writeObject(s)
      out.flush()
      val yourBytes = bos.toByteArray()
      buffer.appendInt(yourBytes.size)
      buffer.appendBytes(yourBytes)
      out.close()
    } catch (e: IOException) {
      e.printStackTrace()
    } finally {
      try {
        bos.close()
      } catch (ex: IOException) {
        ex.printStackTrace()
      }
    }
  }

  override fun decodeFromWire(
    pos: Int,
    buffer: Buffer,
  ): T? {
    // My custom message starting from this *position* of buffer
    var position = pos

    // Length of JSON
    val length = buffer.getInt(position)

    // Jump 4 because getInt() == 4 bytes
    val yourBytes =
      buffer.getBytes(
        4.let {
          position += it
          position
        },
        length.let {
          position += it
          position
        },
      )
    val bis = ByteArrayInputStream(yourBytes)
    try {
      val ois = ObjectInputStream(bis)
      val msg = ois.readObject() as T
      ois.close()
      return msg
    } catch (e: IOException) {
      println("Listen failed " + e.message)
    } catch (e: ClassNotFoundException) {
      println("Listen failed " + e.message)
    } finally {
      try {
        bis.close()
      } catch (e: IOException) {
        e.printStackTrace()
      }
    }
    return null
  }

  override fun transform(customMessage: T): T? {
    // If a message is sent *locally* across the event bus.
    // This example sends message just as is
    return customMessage
  }

  override fun name(): String {
    // Each codec must have a unique name.
    // This is used to identify a codec when sending a message and for unregistering
    // codecs.
    return cls.getSimpleName() + "Codec"
  }

  override fun systemCodecID(): Byte {
    // Always -1
    return -1
  }
}
